jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +170,419 @@ default void clear()
         BufferUtil.clear(getByteBuffer());
     }
 
+    default int append(byte[] bytes, int offset, int length)
+    {
+        ByteBuffer to = getByteBuffer();
+
+        int pos = BufferUtil.flipToFill(to);
+        try
+        {
+            length = Math.min(length, to.remaining());
+            to.put(bytes, offset, length);
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(to, pos);
+        }
+        return length;
+    }
+
+    default void append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+    }
+
+    default boolean append(ByteBuffer bytes, Runnable releaser)
+    {
+        append(bytes);
+        if (bytes.hasRemaining())
+            return false;
+        releaser.run();
+        return true;
+    }
+
+    default void append(RetainableByteBuffer bytes)
+    {
+        bytes.writeTo(this);
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        toInfillMode.put(slice);
+    }
+
+    default boolean writeTo(ByteBuffer to)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
```

Comment:
Why not "consuming" the buffer instead of slicing it?

Reply:
I agree.  We should have all these methods consume.  If  you don't want to consume, then the caller can slice.
However, we may need a slice method on the RetainableByteBuffer API... then it gets a little confusing if that should be a retain/release or not.